make/modules/java.desktop/Lib.gmk

```diff
@@ -34,12 +34,14 @@ $(call FillFindCache, $(wildcard $(TOPDIR)/src/java.desktop/*/native))
 ################################################################################
 # Create the AWT/2D libraries
 
-include lib/Awt2dLibraries.gmk
+ifeq ($(call isTargetOs, android ios), false)
+  include lib/Awt2dLibraries.gmk
```

Comment:
This change seems too extreme.   You'll have access to the desktop module but any use of the java code in this module will fail since libawt will be missing.  The real fix is to configure the sources and makefiles so the entire code path java.desktop -> native libraries is available but only for the headless toolkit.  The libawt_lwawt should not be included.  This may be difficult since the fontmanager is dependent this native code.  

If this is too difficult, you should just go back and not support the java.desktop module like you had in your first PR version.   Although, every time you add a change like this which moves away from full JRE JCK compatibility, you make it more difficult to upstream these changes.

Reply:
It would be possible to fix configure so mobile always use headless mode. From my PoV, it seems like the natural approach. Let me know if you need help with fixing that.

Reply:
The problem with awt_headless is that this depends on X11 (which we don't have on iOS/Android).
I'm not sure what the best approach is. It would need lots of changes to make that work on mobile.
Any help is much appreciated.

Reply:
Now that you mention it, I do remember some fuss long time ago about headless requiring X11. Hmmm. That means that any attempt to use headless is definitely out of bounds for this patch. Maybe the best way for this patch is then to exclude the java.desktop module for iOS and Android in Modules.gmk.

But I'll look into headless once more to see if I can figure out if it *really* needs X11 or if it's just some sloppy programming.

Reply:
Good news! It turned out that java.desktop can build just fine without X11 if you just want headless. 

A quick and dirty hack to confirm that this works for you too:
```
diff --git a/make/autoconf/libraries.m4 b/make/autoconf/libraries.m4
index 5120918aed2..42a641c3228 100644
--- a/make/autoconf/libraries.m4
+++ b/make/autoconf/libraries.m4
@@ -46,7 +46,7 @@ AC_DEFUN_ONCE([LIB_DETERMINE_DEPENDENCIES],
   else
     # All other instances need X11, even if building headless only, libawt still
     # needs X11 headers.
-    NEEDS_LIB_X11=true
+    NEEDS_LIB_X11=false
   fi
 
   # Check if fontconfig is needed
```

I'll open a separate bug to get a fix in mainline so that X11 is not required for headless builds.

Reply:
See. https://bugs.openjdk.java.net/browse/JDK-8255785

Reply:
Thanks for the suggestion.
`NEEDS_LIB_X11` was already false for ios/android (same file, in the if clause) but there were a few places where an X11 header was still required. I could easily fix those, but there are more issues when building libawt.a. For example, font.c relies on fontconfig/fontconfig.h which we don't have on android.
Actually, the whole font handling is something that is already done in JavaFX (in OpenJFX), and for Android we leverage freetype. 
I'm a bit puzzled, as I think it would be a huge overkill to somehow port the font code from OpenJDK to android, but never using it as it is already done in OpenJFX. 
Another area is the dependency on X11 in the java2d/opengl/J2D_GL files.

To me, it seems the main issue is that I don't think this code belongs in the core of OpenJDK.

Reply:
Fine. If you don't need or want java.desktop, then I suggest you exclude that module for iOS/Android.

I still agree with Bob that keeping java.desktop but not building the native libs sounds problematic, since it will make stuff fail surprisingly at runtime.

If you need part, but not all, of java.desktop, then for this patch I don't really care, but if so, that is something that needs to be discussed seriously on how to handle before we can start thinking about upstreaming mobile into mainline.

Reply:
Also, I'm a bit curious of what other places the X11 headers were needed?

Reply:
Maybe try add missing includes before exclude?

Reply:
I'd like to include as much as possible, but if that results in ending up complete implementations between 
#ifndef IOS 
#endif 
statements, I worry that we are moving away from the platform-independent idea as well.

The GLX code is a problem since glx.h has an include to X11/Xlib.h and X11/Xutil.h and it really depends on what is defined there. That header is included in e.g. GLXGraphicsConfig.c (via GLXGraphicsConfig.h)

There is a check on HEADLESS in that file, but it's after the includes. I can move that #ifdef before the includes, but that basically rules out the whole file, and then we are closer to the scenario were everything is included, but all implementations are removed via #ifndef IOS (or #ifndef ANDROID)

I ran into a small build issue, where the -DHEADLESS was not passed to LIBAWT_CFLAGS but that is an easy fix.

Reply:
I think the include of GLXGraphicsConfig.h should really be inside the ifdefs. It does not make sense as it is now. If you look at the GLX code, it is basically split in two, a "dummy" implementation for headless which just does:
```
JNIEXPORT jint JNICALL
Java_sun_java2d_opengl_GLXGraphicsConfig_getOGLCapabilities(JNIEnv *env, jclass glxgc, jlong configInfo) {
    return CAPS_EMPTY;
}
```

and the real implementation for non-headless.

But this got me worried about my fix of building headless without X11. I wonder how I could compile this without X11, or if the compiler dragged up X11 from a local installation without me noticing it. :(

Reply:
I fixed the issues, and with the latest commit, the awt libs are built for ios/android.

Reply:
Cool! :)